docs/spec.md

```diff
@@ -2346,23 +2356,18 @@ More formally, `result[result_index] = dot_product`, where:
 * `transposed_rhs = transpose(rhs, rhs_batching_dimensions + rhs_result_dimensions + rhs_contracting_dimensions)`.
 * `transposed_rhs_slice = slice(transposed_rhs, result_batching_index + result_rhs_index + [:, ..., :])`.
 * `reshaped_rhs_slice = reshape(transposed_rhs_slice, dims(rhs, rhs_contracting_dimensions))`.
-* For non-quantized types:
-  * `dot_product = reduce(
-      inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)],
-      init_values=[constant(0, element_type(result))],
-      dimensions=range(size(lhs_contracting_dimensions)),
-      body=lambda x, y: add(x, y))`.
-* For quantized types:
-  * `integer_dot_product = reduce(
-      inputs=[multiply((reshaped_lhs_slice - zero_point(reshaped_lhs_slice)),
-                       (reshaped_rhs_slice - zero_point(reshaped_rhs_slice))],
-      init_values=[constant(0, storage_type(result)],
-      dimensions=range(size(lhs_contracting_dimensions)),
-      body=lambda x, y: add(x, y))`.
-  * `rounded_dot_product = round_nearest_even(integer_dot_product * (scale(reshaped_lhs_slice) * scale(reshape_rhs_slice) / scale(result)))`.
-  * `dot_product = clamp(storage_min(result), rounded_dot_product + zero_point(result), storage_max(result))`.
+* `dot_product = reduce(
+    inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)],
+    init_values=[constant(0, element_type(result))],
+    dimensions=range(size(lhs_contracting_dimensions)),
+    body=lambda x, y: add(x, y))`.
 <!-- markdownlint-enable line-length -->
 
+For quantized types, performs `dequantize_op_quantize(
```

Comment:
I’d like to check which axes are permitted to be per-axis quantized for dot_general. Convolution states in (C32) and (C33) that per-axis quantization_dimension is an output feature dimension (which makes sense as these are not summed over by the convolution operation). For dot_general, there does not seem to be corresponding constraints. For example, I’d expect a constraint after (C18) that if is_per_axis_quantized(rhs) then quantization_dimension is one of the rhs_result_dimensions. In particular, that it is not one of rhs_contracting_dimensions. Similar constraints also for the result tensor.

Reply:
Thanks for pointing out this important aspect!

Just a bit of background: One of the main reasons per-axis quantization is added in dot_general is because of how dot_general is used in the specification of convolution. I acknowledge that is this not an ideal reason for supporting  per-axis quantization for an op and we identified it as a problem in #1615. As mentioned in the ticket it seems that currently TFLite  does not support per-axis scheme for this op. IMO, that is one reason we skip detailing the quantization behavior for the op is isolation. In any case, I agree with you that the relevant details like `quantized dimension` are missing from the `dot_general` op spec and that is something we should to address if we are allowing the `rhs` and `result` to be per-axis quantized.

The current PR  allows both `convolution` and  `dot_general` op to be defined in terms of floating-point semantics. With that the `convolution` op specification will use the floating-point behavior of dot_general (as opposed to using dot_general op with quantized types). The immediate benefit is that  dot_general does not need to be per-axis quantized just because it is used in the specification of `convolution`. There must be a better reason for the support.

Also, your proposal about quantization dimension to not include the contracting dimensions makes sense. 

I propose that we address it as part of https://github.com/openxla/stablehlo/issues/1615. 

Please let me know your opinion?

Reply:
Thanks @sdasgup3 for explaining the background and the link to #1615.

Since it is still an open question if dot_general needs per-axis quantization, and it is not required for the convolution specification any more, I agree with your comment that the dot_general specification should be per-tensor quantized for now to avoid confusion. So, label (I2) below (line 2394) defining rhs type and result type (line 2405) should be updated to “tensor or per-tensor quantized tensor”?

As you suggest there could be a further update, if needed, when #1615 is resolved.

Reply:
Thanks @dominicsymes 
Makes sense for the `dot_general` to have per-tensor quantization scheme for now. Please take a look at the updated spec.

Reply:
Thanks for updating. Looks good to me.